### PR TITLE
ieee802154: provide general header build and read functions

### PR DESCRIPTION
**Pull request by authmillenon**
First step in a direction to unify IEEE 802.15.4 drivers similar as Ethernet drivers are (and move them to netdev2 in the process).

### CODE REVIEW

**Comment by authmillenon** [comment]
Maybe @A-Paul wants to look over the unittests, too. But be aware that I wrote them from the stand-point that one at least knows the make-up of a IEEE 802.15.4 frame header (little-endian, so first byte/bit in table is least significant), which should be expected if you write a driver for a network device for that protocol ;-):

| Octets: 2 | 1 | 0 or 2 | 0, 2, or 8 | 0 or 2 | 0, 2, or 8 |
| :----------: | :-: | :-----: | :-----------: | :------: | :----------: |
| FCF | Seq | dst_pan | dst_addr | src_pan | src_addr |

where FCF is a bitfield of the following format

| Bits: 0-2 | 3  | 4  | 5  | 6  | 7-9 | 10-11 | 12-13 | 14-15 |
| :---------: | :-: | :-: | :-: | :-: | :--: | :------: | :------: | :------: |
| Frame type | Security Enabled | Frame Pending | ACK REQ | PAN comp | Reserved | dst_mode | version | src_mode |

For the specific values of the FCF fields please see the `ieee802154.h`

**Comment by authmillenon** [comment]
Rebased and squashed

**Comment by authmillenon** [comment]
@aeneby https://github.com/RIOT-OS/RIOT/pull/4645#issuecomment-174415586: if you want those you need to provide them... My knowledge about IEEE802.15.4 is not sufficient enough for that :/

**Comment by aeneby** [comment]
@authmillenon something like:

```C
#define IEEE802154_MIN_FREQ         2405 /**< Min. frequency (2405 MHz) */
#define IEEE802154_MAX_FREQ         2480 /**< Max. frequency (2480 MHz) */

#define IEEE802154_MIN_CHANNEL      11   /**< Min. channel (2405 MHz) */
#define IEEE802154_MAX_CHANNEL      26   /**< Max. channel (2480 MHz) */
#define IEEE802154_CHANNEL_SPACING  5    /**< Channel spacing in MHz  */
```

And if you want conversion macros:

```C
#define IEEE802154_CHAN2FREQ(chan)  ( IEEE802154_MIN_FREQ + ((chan) - IEEE802154_MIN_CHANNEL) * IEEE802154_CHANNEL_SPACING )
#define IEEE802154_FREQ2CHAN(freq)  ( IEEE802154_MIN_CHANNEL + ((freq) - IEEE802154_MIN_FREQ) / IEEE802154_CHANNEL_SPACING )
```

**Comment by authmillenon** [comment]
Isn't this excluding Sub-GHz 802.15.4 radios?

**Comment by gebart** [comment]
@authmillenon @aeneby 
These are some other bands which we need to support as well:

780-786 MHz (China)
868.3 MHz (EU)
906-924 MHz (North America)
Ultra wide band:
249.6 MHz to 749.6 MHz
3.1 GHz to 4.8 GHz
6.0 GHz to 10.6 GHz

**Comment by gebart** [comment]
And something around 950 MHz as well

**Comment by authmillenon** [comment]
Sorry, but that's the point where I say, that this is way to complicated for a "simple" helper macro ;-)

**Comment by aeneby** [comment]
Alright, I guess I'll have to read the spec regarding the other bands too. I still think it's a good idea to have these defined eventually, though. Otherwise they have to be defined per driver.

**Comment by haukepetersen** [comment]
@authmillenon said offline, that I should write: 'this should be simlified' -> so, 'this should be simplified' :-)

**Comment by authmillenon** [comment]
Well I just quoted you ad.lib. ;-)

**Comment by authmillenon** [comment]
Reworked based on discussion at H&A with @haukepetersen (but did not adapt all unittests yet)

**Comment by authmillenon** [comment]
Finished and squashed

**Comment by authmillenon** [comment]
Rebased and squashed again.

**Comment by authmillenon** [comment]
Addressed @thomaseichinger's comments in https://github.com/RIOT-OS/RIOT/pull/4645#discussion_r51443358 and https://github.com/RIOT-OS/RIOT/pull/4645#discussion_r51443436

**Comment by authmillenon** [comment]
Someone care to review?

**Comment by authmillenon** [comment]
@OlegHahm you maybe since you already did some heavy reviewing in the dependent PRs?

**Comment by OlegHahm** [comment]
I didn't do any review, I just tested.

**Comment by authmillenon** [comment]
Rebased to master and squashed.

**Comment by kaspar030** [comment]
@authmillenon I'd like to review. Will go through it in the next days.

**Comment by authmillenon** [comment]
Rebased to current master.

File: sys/include/net/ieee802154.h [hunk 1]
```diff
@@ -43,28 +55,146 @@ extern "C" {
...
+#define IEEE802154_FCF_FRAME_PEND           (0x10)  /**< follow-up frame is pending */
+#define IEEE802154_FCF_ACK_REQ              (0x20)  /**< acknowledgement requested from receiver */
+#define IEEE802154_FCF_PAN_COMP             (0x40)  /**< compress source PAN ID */
 
 #define IEEE802154_FCF_DST_ADDR_MASK        (0x0c)
-#define IEEE802154_FCF_DST_ADDR_VOID        (0x00)
-#define IEEE802154_FCF_DST_ADDR_SHORT       (0x08)
-#define IEEE802154_FCF_DST_ADDR_LONG        (0x0c)
+#define IEEE802154_FCF_DST_ADDR_VOID        (0x00)  /**< no destination address */
+#define IEEE802154_FCF_DST_ADDR_RESV        (0x04)
```
Comment (OlegHahm) [inline sys/include/net/ieee802154.h]: Do we need a constant for a reserved field? If yes, why not document it with `/**< reserved */`?

**Comment by OlegHahm** [comment]
> @authmillenon I'd like to review. Will go through it in the next days.

@kaspar030, found time?

File: sys/include/net/ieee802154.h (hunk 1 above)
Comment (authmillenon) [inline sys/include/net/ieee802154.h]: I think I used it in a previous iteration of the rework. Also I used it in the tests to generate the corner case "User sets wrong flags". Will add doc line.

Comment (authmillenon) [inline sys/include/net/ieee802154.h]: Addressed

**Comment by authmillenon** [comment]
@kaspar030 can I have a drink to celebrate tonight too? :P

**Comment by haukepetersen** [comment]
Looking through this I can't find anything to complain about, so ACK from my side (once Travis agrees of course).

**Comment by authmillenon** [comment]
Rebased and squashed.

**Comment by kaspar030** [comment]
ACK from my side, too. This doesn't touch any existing code, and RIOT CI looks good, so feel free to merge!

**Comment by authmillenon** [comment]
I thought we say we prefer still Travis for this week.

**Comment by kaspar030** [comment]
I thought we take RIOT CI's decisions with a grain of salt and actually take a look at it's result.
This one looks fine. Your call.

**Comment by authmillenon** [comment]
Travis is done now to, so merge ;-P

**Pull request merged**